### PR TITLE
feat(profiler): using approximate queries for profiling

### PR DESCRIPTION
**Pull request by treff7es**
- Running approximate distinct count on Redshift, Snowflake and Bigquery
  - Running approximate qunatiles on Bigquery and Redshift
  - Skipping not known column types from profiling
  - Adding Redshift example recipe with all the profiling settings commented out

## Checklist
- [x ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py [hunk 1]
```diff
@@ -97,6 +101,61 @@ class GEProfilerRequest:
     batch_kwargs: dict
 
 
+def get_column_unique_count_patch(self, column):
+    if self.engine.dialect.name.lower() == "redshift":
+        element_values = self.engine.execute(
+            f"SELECT APPROXIMATE count (distinct {column}) FROM {self._table}"
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:107]: I'd like to lean on SQLAlchemy where possible. Would something along these lines work (might need a `sa.column` in there as well)?

```suggestion
            sa.select([f"APPROXIMATE count (distinct {column})"]).select_from(self._table)
```

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py [hunk 2]
```diff
@@ -97,6 +101,61 @@ class GEProfilerRequest:
...
+        return list(quantiles_results)
+
+    except ProgrammingError as pe:
+        self._treat_quantiles_exception(pe)
+        return list()
+
+
+great_expectations.dataset.sqlalchemy_dataset.SqlAlchemyDataset.get_column_unique_count = (
+    get_column_unique_count_patch
+)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:152]: Would it be possible to perform these patches within the `generate_profiles` method, ideally using `unittest.mock.patch`? That way, the impact of importing DataHub would be limited and a bit easier to reason about.

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py [hunk 3]
```diff
@@ -97,6 +101,61 @@ class GEProfilerRequest:
...
+    select_part = ",".join(quantile_queries)
+
+    quantiles_query = f"SELECT {select_part} " f"from {self._table}"
+    try:
+        quantiles_results = self.engine.execute(quantiles_query).fetchone()
+        return list(quantiles_results)
+
+    except ProgrammingError as pe:
+        self._treat_quantiles_exception(pe)
+        return list()
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:147]: We should note with a comment that `_treat_quantiles_exception` will reraise the exception, so the return statement is unreachable. Might also be more clear to use `raise` in here instead of return.

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py [hunk 4]
```diff
@@ -513,7 +573,8 @@ def generate_dataset_profile(  # noqa: C901 (complexity)
                 columns_profiling_queue.append(column_spec)
 
                 self._get_column_type(column_spec, column)
-                self._get_column_cardinality(column_spec, column)
+                if column_spec.type_ != ProfilerDataType.UNKNOWN:
+                    self._get_column_cardinality(column_spec, column)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:577]: Unfortunately `self._get_column_type` can be executed asynchronously, so it's not guaranteed that `column_spec.type_` will actually be set here.

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py [hunk 5]
```diff
@@ -522,6 +583,11 @@ def generate_dataset_profile(  # noqa: C901 (complexity)
         row_count: int = profile.rowCount
 
         for column_spec in columns_profiling_queue:
+            if column_spec.type_ == ProfilerDataType.UNKNOWN:
+                logger.debug(
+                    f"profiling is disabled for {column_spec.column} as it's type is not supported for profiling"
+                )
+                continue
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:590]: Out of curiosity, what was the motivation for this change? Is there a reason we don't want to get nonnull counts or unique counts for columns with unknown types?

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py [hunk 6]
```diff
@@ -97,6 +101,61 @@ class GEProfilerRequest:
...
+def _get_column_quantiles_bigquery_patch(  # type:ignore
+    self, column: str, quantiles: Iterable
+) -> list:
+    quantile_queries = list()
+    for quantile in quantiles:
+        quantile_queries.append(
+            f"approx_quantiles({column}, 100) OFFSET [{quantile * 100}]"
+        )
+
+    select_part = ",".join(quantile_queries)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:138]: once again, I'd rather that we use SQLAlchemy for some of these operations.

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py (hunk 1 above)
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:107]: I'd like to lean on SQLAlchemy where possible. Would something along these lines work?

```suggestion
            sa.select([sa.column(f"APPROXIMATE count (distinct {column})")]).select_from(self._table)
```

Additionally, we may need to modify the `_is_single_row_query_method` in accordance with these changes.

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py [hunk 7]
```diff
@@ -112,10 +171,10 @@ class GEProfilingConfig(ConfigModel):
...
     include_field_median_value: bool = True
     include_field_stddev_value: bool = True
-    include_field_quantiles: bool = True
-    include_field_distinct_value_frequencies: bool = True
-    include_field_histogram: bool = True
-    include_field_sample_values: bool = True
+    include_field_quantiles: bool = False
+    include_field_distinct_value_frequencies: bool = False
+    include_field_histogram: bool = False
+    include_field_sample_values: bool = False
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:177]: It might be nicer to separate these defaults changes into a separate PR. If it's too much of a hassle, don't worry about it though!

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py [hunk 8]
```diff
@@ -536,11 +602,17 @@ def generate_dataset_profile(  # noqa: C901 (complexity)
                 column_profile.nullCount = null_count
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:602]: I think we need to revise the above assert statement. If non_null_count > row_count, then null_count < 0, and we should simply reset it to 0 instead of throwing an assertion error.

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py (hunk 5 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:598]: Sorry, I should have commented about it. There are types where Profiling fails like GEOGRAPHY type in some dbs. 
I wanted to make sure we don't try to profile these type of columns as there column cardinality might break.

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py (hunk 2 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py]: Thanks, I updated it.

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py (hunk 3 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:163]: I added a comment to this

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py (hunk 4 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py]: I think addressed this as well.

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py (hunk 6 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py]: As we discussed I rewrote it with SQLAlchemy

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py (hunk 1 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py]: I'm not sure if _is_single_row_query_method needs to be changed as all the methods returns similar values than before only some inner method got monkey patched.

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py (hunk 7 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py]: Thanks, I will create a separate pr from this.

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py (hunk 8 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:616]: I updated this as well.

File: metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py [hunk 9]
```diff
@@ -704,27 +769,34 @@ def generate_profiles(
...
                 f"Will profile {len(requests)} table(s) with {max_workers} worker(s) - this may take a while"
             )
+            with unittest.mock.patch(
+                "great_expectations.dataset.sqlalchemy_dataset.SqlAlchemyDataset.get_column_unique_count",
+                get_column_unique_count_patch,
+            ):
+                with unittest.mock.patch(
+                    "great_expectations.dataset.sqlalchemy_dataset.SqlAlchemyDataset._get_column_quantiles_bigquery",
+                    _get_column_quantiles_bigquery_patch,
+                ):
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/ge_data_profiler.py:779]: Could you wrap these into the existing `with` statement?

More of an aesthetic thing to avoid all the extra indentation

**Review by hsheth2 — APPROVED** [review]
LGTM other than a small nit!

**Review by shirshanka — APPROVED** [review]
LGTM
